lib/datadog/tracing/trace_operation.rb

```diff
@@ -240,15 +240,27 @@ def build_span(
         end
       end
 
+      # Returns a {TraceSegment} with all finished spans that can be flushed
+      # at invocation time. All other **finished** spans are discarded.
+      #
+      # A span can be flushed if:
+      #
+      # 1. The span has finished, and
+      # 2. Either:
+      #   a. The trace is kept by sampling.
+      #   b. The span is kept by single span sampling.
+      #
+      # Unfinished spans are not affected by this method.
+      #
+      # @return [TraceSegment]
       def flush!
-        finished = finished?
-
-        # Copy out completed spans
-        spans = @spans.dup
-        @spans = []
-
-        # Use them to build a trace
-        build_trace(spans, !finished)
+        if sampled?
+          flush_all_spans!
+        else
+          # Only spans where the span-level sampling overrides
+          # the trace-level sampling can be flushed.
+          flush_single_sampled_spans_only!
```

Comment:
Similar to my feedback regarding `Span` and sampling, I think the same logic may apply here. Perhaps the better place for this kind of responsibility is in the `Flush` classes. They are supposed to know how to extract the "right" spans from an active trace.